src/util/bson/bson.cpp

```diff
@@ -520,106 +523,64 @@ class Parser : public nlohmann::json_sax<json> {
         Type m_type;
 
         union {
-            BsonDocument* document;
-            BsonArray* array;
+            std::unique_ptr<BsonDocument> document;
+            std::unique_ptr<BsonArray> array;
         };
     public:
         ~BsonContainer() noexcept
         {
             if (m_type == DOCUMENT) {
-                delete document;
-                document = nullptr;
+                document.reset();
             } else {
-                delete array;
-                array = nullptr;
+                array.reset();
             }
         }
 
-        BsonContainer(const BsonDocument& v) noexcept : document(new BsonDocument(v)) {
-            m_type = DOCUMENT;
-        }
-        BsonContainer(const BsonArray& v) noexcept : array(new BsonArray(v)) {
-            m_type = ARRAY;
+        BsonContainer(const BsonDocument& v) noexcept :
+            m_type(DOCUMENT),
+            document(new BsonDocument(v)) {
         }
+
+        BsonContainer(const BsonArray& v) noexcept :
+            m_type(ARRAY),
+            array(new BsonArray(v)) {}
+
         BsonContainer(BsonDocument&& v) noexcept :
-            m_type(DOCUMENT), document(new BsonDocument(std::move(v))) {}
+            m_type(DOCUMENT),
+            document(new BsonDocument(std::move(v))) {}
 
-        BsonContainer(BsonArray&& v) noexcept : m_type(ARRAY), array(new BsonArray(std::move(v))) {}
+        BsonContainer(BsonArray&& v) noexcept :
+            m_type(ARRAY),
+            array(new BsonArray(std::move(v))) {}
 
-        explicit operator const BsonDocument&&() const {
+        explicit operator const BsonDocument&() noexcept {
             REALM_ASSERT(is_document());
-            return std::move(*document);
+            return *document.release();
```

Comment:
`release()` will cause a leak because it stops managing the memory, you probably want `get()` or simply `*document`
Same for below instances.

Reply:
moved to just dereferencing document